### PR TITLE
Add faro.receiver component

### LINKED ISSUE #2320: Flow: Add app_agent_receiver integration component(s)
Add one or more components for the `integrations-next` [app_agent_receiver integration](https://grafana.com/docs/agent/next/configuration/integrations/integrations-next/app-agent-receiver-config/).

### PR DESCRIPTION
**Pull request by rfratto**
#### PR Description

This PR adds a component for faro.reciever, an equivalent to the app_agent_receiver integration from static mode. 

#### Which issue(s) this PR fixes

Supersedes #5243 
Closes #2320

#### Notes to the Reviewer

Unfortunately, unlike my other large PRs, this one is **not** very reviewable commit-by-commit. 

This is not 100% a straight port from the existing integration. I have done some minor refactoring to make it possible for this component to be dynamically updateable at runtime. However, the implementation remains mostly the same. 

During my porting, I have noticed bugs in integration, particularly around sourcemaps. I have not fixed these, but I have added TODO comments in the code for them to be resolved down the line. 

Unfortunately, the resulting code is quite complex. Future changes to refactor it to make it easier to read are warranted, but it would likely require more code and would drift further from the original integration. In the spirit of this being a "port," such changes are not included here. 

#### PR Checklist

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated
- [X] Config converters updated (N/A since app_agent_receiver is a v2 integration, and v2 integrations are not supported yet with config conversion.)

### CODE REVIEW

File: docs/sources/flow/reference/components/faro.receiver.md [hunk 1]
```diff
@@ -0,0 +1,267 @@
+---
+aliases:
+- /docs/grafana-cloud/agent/flow/reference/components/faro.receiver/
+- /docs/grafana-cloud/monitor-infrastructure/agent/flow/reference/components/faro.receiver/
+- /docs/grafana-cloud/monitor-infrastructure/integrations/agent/flow/reference/components/faro.receiver/
+canonical: https://grafana.com/docs/agent/latest/flow/reference/components/faro.receiver/
+title: faro.receiver
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/faro.receiver.md:7]:
```suggestion
title: faro.receiver
description: Learn about the faro.receiver
```

We need to add the description: metadata in all files. It doesn't have to be elaborate, just present with some simple text.

File: docs/sources/flow/reference/components/faro.receiver.md [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+`cors_allowed_origins` | `list(string)` | Origins for which cross-origin requests are permitted. | `[]` | no
+`api_key` | `secret` | Optional API key to validate client requests with. | `""` | no
+`max_allowed_payload_size` | `number` | Maximum size (in bytes) for client requests. | `5e6` | no
+
+By default, telemetry data is only accepted from applications on the same local
+network as the browser. To accept telemetry data from a wider set of clients,
+modify the `listen_address` attribute to the IP address of the appropriate
+network interface to use.
+
+The `cors_allowed_origins` argument determines what oriigns browser requests
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/faro.receiver.md:73]:
```suggestion
The `cors_allowed_origins` argument determines what origins browser requests
```

File: docs/sources/flow/reference/components/faro.receiver.md [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+The `rate_limiting` block configures rate limiting for client requests.
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`enabled` | `bool` | Whether to enable rate limiting. | `true` | no
+`rate` | `number` | Rate of allowed requests per second. | `50` | no
+`burst_size` | `number` | Allowed burst size of requests. | `100` | no
+
+Rate limiting functions as a [token bucket algorithm][token-bucket], where
+a bucket has maximum capacity for up to `burst_size` requests, and refills at a
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/faro.receiver.md:96]:
```suggestion
a bucket has a maximum capacity for up to `burst_size` requests and refills at a
```

File: docs/sources/flow/reference/components/faro.receiver.md [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+rate of `rate` per second.
+
+Each HTTP request drains the capacity of the bucket by one. Once the bucket is
+empty, HTTP requests are rejected with a `HTTP 429 Too Many Requests` status
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/faro.receiver.md:100]:
```suggestion
empty, HTTP requests are rejected with an `HTTP 429 Too Many Requests` status
```

File: docs/sources/flow/reference/components/faro.receiver.md [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+code until the bucket has more availble capacity.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/faro.receiver.md:101]:
```suggestion
code until the bucket has more available capacity.
```

File: docs/sources/flow/reference/components/faro.receiver.md [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+file and line information from the original source code.
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`download` | `bool` | Whether to download sourcemaps. | `true` | no
+`download_from_origins` | `list(string)` | Which origins to download sourcemaps from. | `["*"]` | no
+`download_timeout` | `duration` | Timeout when downloading sourcemaps. | `"1s"` | no
+
+When exceptions are sent to the `faro.receiver` component, it can download
+sourcemaps from the web application. This behavior can be disabled by setting
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/faro.receiver.md:122]:
```suggestion
sourcemaps from the web application. You can disable this behavior by setting
```

File: docs/sources/flow/reference/components/faro.receiver.md [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    client {
+        endpoint = "OTLP_ADDRESS"
+    }
+}
+```
+
+Replace the following:
+
+* `NETWORK_ADDRESS`: IP address of the network interface to listen to traffic
+  on. This IP adddress must be reachable by browsers using the web application
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/faro.receiver.md:249]:
```suggestion
  on. This IP address must be reachable by browsers using the web application
```

File: docs/sources/flow/reference/components/faro.receiver.md [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+and forwards it to other components for future processing.
+
+[faro-sdk]: https://github.com/grafana/faro-web-sdk
+
+## Usage
+
+```river
+faro.receiver "LABEL" {
+    output {
+        logs   = [LOGS_RECEIVERS]
```
Comment (tpaschalis) [inline docs/sources/flow/reference/components/faro.receiver.md:22]: Is this an OTel log receiver (like traces below), or a loki.* component?

*edit*: I see this is explained in the `output` block below, nvm.

File: docs/sources/flow/reference/components/faro.receiver.md [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+component. Clients using the [Grafana Faro Web SDK][faro-sdk] forward telemetry
+data to this HTTP server for processing.
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`listen_address` | `string` | Address to listen for HTTP traffic on. | `127.0.0.1` | no
+`listen_port` | `number` | Port to listen for HTTP traffic on. | `12347` | no
+`cors_allowed_origins` | `list(string)` | Origins for which cross-origin requests are permitted. | `[]` | no
+`api_key` | `secret` | Optional API key to validate client requests with. | `""` | no
+`max_allowed_payload_size` | `number` | Maximum size (in bytes) for client requests. | `5e6` | no
```
Comment (tpaschalis) [inline docs/sources/flow/reference/components/faro.receiver.md:66]: nit: In the future we can have this use unit2bytes like we do for other components. Let's just leave a TODO comment for it.

File: docs/sources/flow/reference/components/faro.receiver.md [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+template value, such as `/var/my-app/{{ .Release }}/build`. The template value
+will be replaced with the release value provided by the [Faro Web App SDK][faro-sdk].
+
+### output block
+
+The `output` block specifies where to forward collected logs and traces.
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`logs` | `list(LogsReceiver)` | A list of `loki` components to forward lgos to. | `[]` | no
```
Comment (tpaschalis) [inline docs/sources/flow/reference/components/faro.receiver.md:179]:
```suggestion
`logs` | `list(LogsReceiver)` | A list of `loki` components to forward logs to. | `[]` | no
```

File: component/faro/receiver/internal/payload/utils.go [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+import (
+	"fmt"
+	"sort"
+
+	om "github.com/wk8/go-ordered-map"
+)
+
+// KeyVal is an ordered map of string to interface
+type KeyVal = om.OrderedMap
```
Comment (tpaschalis) [inline component/faro/receiver/internal/payload/utils.go:11]: Will this be a problem if we define two separate faro.receiver components?

**Review by tpaschalis — APPROVED** [review]
LGTM. 

I've done a quick pass through (so not entirely rubber-stamping), nothing pops out other than that package-level variable and the various TODOs we've left, but these can be fixed in due time.

File: component/faro/receiver/internal/payload/utils.go (hunk 11 above)
Comment (rfratto) [reply component/faro/receiver/internal/payload/utils.go:11]: Why would it be? This is a type alias, not a variable :)

Comment (tpaschalis) [reply component/faro/receiver/internal/payload/utils.go:11]: Oops, sorry I needed my second coffee of the day 😅

File: docs/sources/flow/reference/components/faro.receiver.md (hunk 9 above)
Comment (rfratto) [reply docs/sources/flow/reference/components/faro.receiver.md:67]: I'll change that now; shouldn't be too much effort.

**Comment by gavinwoolley** [comment]
Hi, good work on this, I am eagerly awaiting this feature. 
If there was one thing i'd like to see in future, is the capability to work with multi tenants backends from a single receiver.
It requires multiple configs and ports defining for multi tenant currently

**Comment by rfratto** [comment]
> Hi, good work on this, I am eagerly awaiting this feature.
If there was one thing i'd like to see in future, is the capability to work with multi tenants backends from a single receiver.
It requires multiple configs and ports defining for multi tenant currently

Hi, due to how the component works now (where one receiver can forward to multiple components), it should be possible to do a basic form of multi-tenancy through creative use of sending to multiple filtering components (i.e., filter by tenant). This unfortunately requires knowing the set of tenants in advance.

More dynamic support for multitenancy has been discussed a few times, but it's not something we're planning at the moment. If you're interested in seeing that, please feel free to open a feature request :)

**Review by clayton-cornell — APPROVED** [review]

**Pull request merged**